src/main/java/io/confluent/copycat/hdfs/TopicPartitionWriter.java

```diff
@@ -0,0 +1,461 @@
+  private Set<String> appended;
+  private long offset = -1;
+  private Map<String, Long> startOffsets;
+  private Map<String, Long> offsets;
+  private long timeoutMs;
+  private long failureTime;
+
+  private Pattern pattern = Pattern.compile("\\w+_\\d+_\\d+_\\d+");
+
+  public TopicPartitionWriter(String url,
+                              String topicsDir,
+                              String logsDir,
+                              TopicPartition tp,
+                              SinkTaskContext context,
+                              Partitioner partitioner,
+                              RecordWriterProvider writerProvider,
+                              Storage storage,
+                              int flushSize,
+                              long timeoutMs,
+                              Configuration conf,
+                              AvroData avroData) {
+    this.storage = storage;
+    this.writerProvider = writerProvider;
+    this.tp = tp;
+    this.partitioner = partitioner;
+    this.context = context;
+    this.url = url;
+    this.topicsDir = topicsDir;
+    this.flushSize = flushSize;
+    this.timeoutMs = timeoutMs;
+    this.conf = conf;
+    this.avroData = avroData;
+    wal = storage.wal(logsDir, tp);
+    buffer = new LinkedList<>();
+    writers = new HashMap<>();
+    tempfiles = new HashMap<>();
+    appended = new HashSet<>();
+    startOffsets = new HashMap<>();
+    offsets = new HashMap<>();
+    state = State.RECOVERY_STARTED;
+    failureTime = -1;
+  }
+
+  private enum State {
+    RECOVERY_STARTED,
+    RECOVERY_PARTITION_PAUSED,
+    WAL_CREATED,
+    WAL_APPLIED,
+    WAL_TRUNCATED,
+    OFFSET_RESET,
+    WRITE_STARTED,
+    WRITE_PARTITION_PAUSED,
+    SHOULD_ROTATE,
+    TEMP_FILE_CLOSED,
+    WAL_APPENDED,
+    FILE_COMMITTED;
+
+    private static State[] vals = values();
+    public State next() {
+      return vals[(this.ordinal() + 1) % vals.length];
+    }
+  }
+
+  public void recover() {
+    try {
+      switch (state) {
+        case RECOVERY_STARTED:
+          pause();
+          nextState();
+        case RECOVERY_PARTITION_PAUSED:
+          createWAL();
+          nextState();
+        case WAL_CREATED:
+          applyWAL();
+          nextState();
+        case WAL_APPLIED:
+          truncateWAL();
+          nextState();
+        case WAL_TRUNCATED:
+          resetOffsets();
+          nextState();
+        case OFFSET_RESET:
+          resume();
+          nextState();
+          break;
+        default:
+          log.error("{} is not a valid state to perform recovery for topic partition {}.", state, tp);
+      }
+    } catch (CopycatException e) {
+      log.error("Recovery failed.");
+      setRetryTimeout(timeoutMs);
+    }
+  }
+
+  public void write() {
+    long now = System.currentTimeMillis();
+    if (failureTime > 0 && now - failureTime < timeoutMs) {
+      return;
+    }
+    if (state.compareTo(State.WRITE_STARTED) < 0) {
+      recover();
```

Comment:
Just noticed that I think we missed something in the previous patch. When reusing `recover()` here, we need it to return a boolean or something -- if it does fail, then we should not continue to the loop below this, right? I think we should just be returning in that case.